lib/VolSurface/Calibration/Equities.pm

```diff
@@ -0,0 +1,637 @@
+    my @current_contraction;
+
+    my $expansion_function_eval;
+    my $reflected_function_eval;
+    my $contraction_function_eval;
+    my $lowest_point_index = 0;
+
+    my $calcvert = 1;
+    my $counter  = 0;
+
+    while (1) {
+        if ($calcvert == 1) {
+            $calcvert = 0;
+            for ($j = 0; $j <= $num_of_points; $j++) {
+                if ($j != $lowest_point_index) {
+                    for ($i = 0; $i < $num_of_points; $i++) {
+                        $current_reflection[$i] = $simplex[$j][$i];
+                    }
+                    $function_eval[$j] = $self->function_to_optimize(\@current_reflection);
+                }
+            }
+
+        }
+        my @sorted_eval = sort { $a <=> $b } @function_eval;
+        $lowest_point_index = first_index { $_ == $sorted_eval[0] } @function_eval;
+        my $highest_point_index = first_index { $_ == $sorted_eval[-1] } @function_eval;
+        my @indexes = indexes { $_ == $sorted_eval[-2] } @function_eval;
+
+        my @simplex_centroid = _calculate_simplex_centroid(\@simplex, $num_of_points, $num_of_var, $highest_point_index);
+
+        $highest_value = $function_eval[$highest_point_index];
+
+        my $convtol = $intol * ($starting_value + $intol);
+
+        last
+            if ($function_eval[$lowest_point_index] < $intol
+            or ($function_eval[$highest_point_index] < ($function_eval[$lowest_point_index] + $convtol))
+            or $counter > $self->_max_iteration);
+
+        #imple reflection of the highest point
+        @current_reflection =
+            map { $simplex_centroid[$_] + 1 * ($simplex_centroid[$_] - $simplex[$highest_point_index][$_]) } (0 .. $num_of_var - 1);
+        my $reflected_function_eval = $self->function_to_optimize(\@current_reflection);
+        if ($reflected_function_eval < $function_eval[$lowest_point_index]) {
+
+#Do simple expansion or in other words look up a little further in this direction
+            @current_expansion = map { $simplex_centroid[$_] + 2 * ($current_reflection[$_] - $simplex_centroid[$_]) } (0 .. $num_of_var - 1);
+
+            @simplex_centroid =
+                map { $current_reflection[$_] } (0 .. $num_of_var - 1);
+            $expansion_function_eval = $self->function_to_optimize(\@current_expansion);
+            if ($expansion_function_eval < $reflected_function_eval) {
+
+                #replace highest point with expansion point
+                #assing function value to highest point
+                @{$simplex[$highest_point_index]} =
+                    map { $current_expansion[$_] } (0 .. $num_of_var - 1);
+                $function_eval[$highest_point_index] = $expansion_function_eval;
+            } else {
+
+                #replace highest point with reflected point
+                #assign highest value to reflection point
+                @{$simplex[$highest_point_index]} =
+                    map { $simplex_centroid[$_] } (0 .. $num_of_var - 1);
```

Comment:
This note is related to the *see below* in line 487.

`@simplex_centroid` at this point contains the same data as `@current_reflection`. So why not

```perl
@{$simplex[$highest_point_index]} = @current_reflection;
```

Then the copy operation in line 487 is not necessary at all.